src/WebJobs.Script.WebHost/Models/ContainerFunctionExecutionActivity.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Models
+{
+    public class ContainerFunctionExecutionActivity
+    {
+        public ContainerFunctionExecutionActivity(DateTime eventTime, string functionName,
+            ExecutionStage executionStage, string triggerType, bool success)
+        {
+            EventTime = eventTime;
+            FunctionName = functionName;
+            ExecutionStage = executionStage;
+            TriggerType = triggerType;
+            Success = success;
+        }
+
+        public DateTime EventTime { get; set; }
+
+        public string FunctionName { get; }
+
+        public ExecutionStage ExecutionStage { get; }
+
+        public string TriggerType { get; }
+
+        public bool Success { get; }
+
+        public override string ToString()
+        {
+            return $"{FunctionName}:{ExecutionStage}-{Success}";
+        }
+
+        protected bool Equals(ContainerFunctionExecutionActivity other)
+        {
+            return string.Equals(FunctionName, other.FunctionName) && ExecutionStage == other.ExecutionStage &&
+                   string.Equals(TriggerType, other.TriggerType) && Success == other.Success;
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj))
+            {
+                return false;
+            }
+
+            if (ReferenceEquals(this, obj))
+            {
+                return true;
+            }
+
+            if (obj.GetType() != this.GetType())
+            {
+                return false;
+            }
+            return Equals((ContainerFunctionExecutionActivity)obj);
+        }
+
+        // ignores EventTime intentionally
```

Comment:
So you're relying on this + use of HashSet to de-dupe per Function so you only publish the "last" activity status per function? That requires that you hold all of these in memory unnecessarily prior to the de-dupe/publish. Wouldn't it be simpler to just have a dictionary of Function activities in the publisher (keyed by name) and just update the record for each when activities are added? That way we just keep what we need. Also, the logic is then less subtle.

Reply:
Added logic to dedup while adding to the list instead of when removing it.